pgdrive/scene_manager/traffic_manager.py

```diff
@@ -0,0 +1,379 @@
+        :return:
+        """
+        vehicles = [
+            v for v in self.vehicles
+            if norm((v.position - vehicle.position)[0], (v.position - vehicle.position)[1]) < distance
+            and v is not vehicle and (see_behind or -2 * vehicle.LENGTH < vehicle.lane_distance_to(v))
+        ]
+
+        vehicles = sorted(vehicles, key=lambda v: abs(vehicle.lane_distance_to(v)))
+        if count:
+            vehicles = vehicles[:count]
+        return vehicles
+
+    def neighbour_vehicles(self, vehicle, lane_index: Tuple = None) -> Tuple:
+        """
+        Find the preceding and following vehicles of a given vehicle.
+
+        :param vehicle: the vehicle whose neighbours must be found
+        :param lane_index: the lane on which to look for preceding and following vehicles.
+                     It doesn't have to be the current vehicle lane but can also be another lane, in which case the
+                     vehicle is projected on it considering its local coordinates in the lane.
+        :return: its preceding vehicle, its following vehicle
+        """
+        lane_index = lane_index or vehicle.lane_index
+        if not lane_index:
+            return None, None
+        lane = self.map.road_network.get_lane(lane_index)
+        s = self.map.road_network.get_lane(lane_index).local_coordinates(vehicle.position)[0]
+        s_front = s_rear = None
+        v_front = v_rear = None
+        for v in self.vehicles:
+            if norm(v.position[0] - vehicle.position[0], v.position[1] - vehicle.position[1]) > 100:
+                # coarse filter
+                continue
+            if v is not vehicle:
+                s_v, lat_v = lane.local_coordinates(v.position)
+                if not lane.on_lane(v.position, s_v, lat_v, margin=1):
+                    continue
+                if s <= s_v and (s_front is None or s_v <= s_front):
+                    s_front = s_v
+                    v_front = v
+                if s_v < s and (s_rear is None or s_v > s_rear):
+                    s_rear = s_v
+                    v_rear = v
+        return v_front, v_rear
+
+    def destroy(self, pg_world: PGWorld) -> None:
+        """
+        Destory func, release resource
+        :param pg_world: World
+        :return: None
+        """
+        self.clear_traffic(pg_world)
+        # current map
+        self.map = None
+
+        # traffic vehicle list
+        self.ego_vehicle = None
+        self.vehicles = None
+        self.traffic_vehicles = None
+        self.block_triggered_vehicles = None
+
+        # traffic property
+        self.mode = None
+        self.random_traffic = None
+        self.density = None
+
+        # control randomness of traffic
+        self.random_seed = None
+        self.np_random = None
+
+    def __del__(self):
+        logging.debug("{} is destroyed".format(self.__class__.__name__))
+
+    def __repr__(self):
+        return self.vehicles.__repr__()
```

Comment:
I can't see the diff between this file and the old scene manager by git. Could you tell me what major changes have been made in this file?

Reply:
The traffic manager part in scene_manager is moved to the new class ```class TrafficManager``` , and the Record&Replay system still remains in scene_manager to access/record/restore all runtime elements. 
A new traffic mode named **hybrid** is newly added to traffic manager, which means the triggered vehicles will be reborn instead of removing, and I have tested all scripts. The scene_manager will be a global runtime manager in the future, managing traffic light/pedestrian/objects and so on.